ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/layered/AbstractTransactionsLayer.java

```diff
@@ -0,0 +1,593 @@
+      final PendingTransaction pendingTransaction);
+
+  @Override
+  public PendingTransaction promote(final Address sender, final long nonce) {
+    final var senderTxs = txsBySender.get(sender);
+    if (senderTxs != null) {
+      long expectedNonce = nonce + 1;
+      if (senderTxs.firstKey() == expectedNonce) {
+        final PendingTransaction promotedTx = senderTxs.pollFirstEntry().getValue();
+        processRemove(senderTxs, promotedTx.getTransaction(), PROMOTED);
+        metrics.incrementRemoved(promotedTx.isReceivedFromLocalSource(), "promoted", name());
+
+        if (senderTxs.isEmpty()) {
+          txsBySender.remove(sender);
+        }
+        return promotedTx;
+      }
+    }
+    return nextLayer.promote(sender, nonce);
+  }
+
+  private TransactionAddedResult addToNextLayer(
+      final PendingTransaction pendingTransaction, final int distance) {
+    return addToNextLayer(
+        txsBySender.getOrDefault(pendingTransaction.getSender(), EMPTY_SENDER_TXS),
+        pendingTransaction,
+        distance);
+  }
+
+  private TransactionAddedResult addToNextLayer(
+      final NavigableMap<Long, PendingTransaction> senderTxs,
+      final PendingTransaction pendingTransaction,
+      final int distance) {
+    final int nextLayerDistance;
+    if (senderTxs.isEmpty()) {
+      nextLayerDistance = distance;
+    } else {
+      nextLayerDistance = (int) (pendingTransaction.getNonce() - (senderTxs.lastKey() + 1));
+    }
+    return nextLayer.add(pendingTransaction, nextLayerDistance);
+  }
+
+  private void processAdded(final PendingTransaction addedTx) {
+    pendingTransactions.put(addedTx.getHash(), addedTx);
+    final var senderTxs = txsBySender.computeIfAbsent(addedTx.getSender(), s -> new TreeMap<>());
+    senderTxs.put(addedTx.getNonce(), addedTx);
+    increaseSpaceUsed(addedTx);
+    metrics.incrementAdded(addedTx.isReceivedFromLocalSource(), name());
+    internalAdd(senderTxs, addedTx);
+  }
+
+  protected abstract void internalAdd(
+      final NavigableMap<Long, PendingTransaction> senderTxs, final PendingTransaction addedTx);
+
+  protected abstract int maxTransactionsNumber();
+
+  private void evict(final long spaceToFree, final int txsToEvict) {
+    final var evictableTx = getEvictable();
+    if (evictableTx != null) {
+      final var lessReadySender = evictableTx.getSender();
+      final var lessReadySenderTxs = txsBySender.get(lessReadySender);
+
+      long evictedSize = 0;
+      int evictedCount = 0;
+      PendingTransaction lastTx;
+      // lastTx must never be null, because the sender have at least the lessReadyTx
+      while ((evictedSize < spaceToFree || txsToEvict > evictedCount)
+          && !lessReadySenderTxs.isEmpty()) {
+        lastTx = lessReadySenderTxs.pollLastEntry().getValue();
+        processEvict(lessReadySenderTxs, lastTx);
+        ++evictedCount;
+        evictedSize += lastTx.memorySize();
+        // evicted can always be added to the next layer
```

Comment:
Why is this true when evicting from Prioritised? What if Ready space is full?

Reply:
probably the wording is not clear, but since txs in a layer are in average better than the following layer, it make sense to says it will always find a place in Ready, and if Ready is full after adding this tx, some other will be evicted and so on.

The rule is add first then evict if full